ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+
+func createWriterForCheckpointSummary(outputDir string, outputFile string, logger *zerolog.Logger) (io.WriteCloser, error) {
+	fullPath := filePathHeader(outputDir, outputFile)
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file already exists at %v", fullPath)
+	}
+
+	return createClosableWriter(outputDir, logger, outputFile, fullPath)
+}
+
+// 17th part file contains:
+// 1. checkpoint version TODO
+// 2. checkpoint file part index TODO
+// 3. subtrieNodeCount TODO
+// 4. top level nodes
+// 5. trie roots
+// 6. node count
+// 7. trie count
+// 6. checksum
+func storeTopLevelNodesAndTrieRoots(
+	tries []*trie.MTrie,
+	subTrieRootIndices map[*node.Node]uint64,
+	subTriesNodeCount uint64,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) (uint32, error) {
+	// the remaining nodes and data will be stored into the same file
+	closable, err := createWriterForTopTries(outputDir, outputFile, logger)
+	if err != nil {
+		return 0, fmt.Errorf("could not create writer for top tries: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	writer := NewCRC32Writer(closable)
+
+	topLevelNodeIndices, topLevelNodesCount, err := storeTopLevelNodes(
+		tries,
+		subTrieRootIndices,
+		subTriesNodeCount+1, // the counter is 1 more than the node count, because the first item is nil
+		writer)
+
+	if err != nil {
+		return 0, fmt.Errorf("could not store top level nodes: %w", err)
+	}
+
+	logger.Info().Msgf("top level nodes have been stored. top level node count: %v", topLevelNodesCount)
+
+	err = storeRootNodes(tries, topLevelNodeIndices, flattener.EncodeTrie, writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store top level nodes: %w", err)
+	}
+
+	err = storeFooter(topLevelNodesCount, uint16(len(tries)), writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store footer: %w", err)
+	}
+
+	// write checksum to the end of the file
+	checksum := writer.Crc32()
+	_, err = writer.Write(encodeCRC32Sum(checksum))
+	if err != nil {
+		return 0, fmt.Errorf("cannot write CRC32 checksum to top level part file: %w", err)
+	}
+	return checksum, nil
+}
+
+func createSubTrieRoots(tries []*trie.MTrie) [subtrieCount][]*node.Node {
+	var subtrieRoots [subtrieCount][]*node.Node
+	for i := 0; i < len(subtrieRoots); i++ {
+		subtrieRoots[i] = make([]*node.Node, len(tries))
+	}
+
+	for trieIndex, t := range tries {
+		// subtries is an array with subtrieCount trie nodes
+		// in breadth-first order at subtrieLevel of the trie `t`
+		subtries := getNodesAtLevel(t.RootNode(), subtrieLevel)
+		for subtrieIndex, subtrieRoot := range subtries {
+			subtrieRoots[subtrieIndex][trieIndex] = subtrieRoot
+		}
+	}
+	return subtrieRoots
+}
+
+func estimateSubtrieNodeCount(tries []*trie.MTrie) int {
+	if len(tries) == 0 {
+		return 0
+	}
+	estimatedTrieNodeCount := 2*int(tries[0].AllocatedRegCount()) - 1
```

Comment:
I would suggest we take the last trie and use the allocatedRegCount from there considering its most likely have more registers than the trie with 0 index.